Added all necessary UI Components

PplPrinceDalsaniya opened the pull request:
I have added the Components for my RecipeApp. Here is the hierarchy of components.
- app
  - header [has navbar]
  - recipes
     - recipe-details [click any recipe in list, it will show in this section]
     - recipe-list [list containing all recipes]
  - shared
    - ingredient
  - shopping-list [bottom last section containing all ingredients in your shopping list]
    - shopping-edit [top section with all edit functionalities]
    
![](https://user-images.githubusercontent.com/97096354/155457103-0459119d-2ebd-46eb-bc92-6a7ed2c0c102.png)

PplPrinceDalsaniya committed:
Added all necessary UI Components

PplPrinceDalsaniya committed:
Made navbar links Recipes and ShoppingList working

SavvasPalashJain committed:
Clicks to recipelist and shows that recipe details

PplPrinceDalsaniya committed:
Made Add ingredient button working

PplPrinceDalsaniya committed:
Managing RecipesList from Recipe Service

PplPrinceDalsaniya committed:
Handling recipe-item click by recipeService.

PplPrinceDalsaniya committed:
Adding Ingredients by ShoppingListService

PplPrinceDalsaniya committed:
Showing Ingredients in RecipeDetails section

PplPrinceDalsaniya committed:
Add ingredients of recipe to shoppingList from ingredientDetails

PplPrinceDalsaniya committed:
Setting Up Routes.

- Added router-outlet instead of if-else logic
- outsourcing of routes from app-routing.module.ts

PplPrinceDalsaniya committed:
Adding Navigation

- Removed href and click event listeners
- Added routerLink for navigating to the section of navbar

PplPrinceDalsaniya committed:
Added routerLinkActive

- to show the current page, in navbar.

PplPrinceDalsaniya committed:
Fixed Page Reload issue

- Removed all href="#" in 'a' tags

PplPrinceDalsaniya committed:
Started Adding Child Routes

- Added Child route for page where no recipe is selcted.
- For that we also created a component recipe-start.
- removed if-else logic for showing a recipe in recipe detail section.

PplPrinceDalsaniya committed:
Configuring Route Parameter

- added child route for :id
- subscribed to change of params, so, whenever clicks different id it should load that recipe detail
- added getRecipe methode in recipe service to get recipe by ID

PplPrinceDalsaniya committed:
Passing Dynamic Id to route.

- Passing index from recipe-list to recipe-item component by proprty binding.
- Added router link to recipe-item.

PplPrinceDalsaniya committed:
Added RecipeEdit Component and set routes for Editing

PplPrinceDalsaniya committed:
Added Routes for Edit page and NewRecipe button

PplJDHirapara commented on src/app/app.component.ts:
```diff
@@ -6,5 +6,9 @@ import { Component } from '@angular/core';
-  title = 'recipeApp';
+  loadedSection = 'recipes';
+
+  onSelection(section: string) {
```
Can you update the method name to have it as more descriptive like oSectionSelection may be?

PplJDHirapara commented on src/app/header/header.component.html:
```diff
@@ -0,0 +1,30 @@
+<nav class="navbar navbar-expand-lg navbar-light bg-light mb-3">
+  <div class="container-fluid">
+    <a class="navbar-brand" href="#">Recipe Book</a>
```
Take all static values of all html files from en.json using translate module

PplPrinceDalsaniya replied:
I am not sure, if I have got the question right or not. But what I think you are asking here is to implement translate module, so that we can translate it to any language in browser. And those fixed texts like headings and titles should be stored in "en.json" and I should get it from there instead of directly writing them here. 

Please correct me if I got this wrong.

PplJDHirapara commented on src/app/recipes/recipe-edit/recipe-edit.component.ts:
```diff
@@ -0,0 +1,15 @@
+
+  constructor() { }
+
+  ngOnInit(): void {
```
Remove unused constructor and method from whole app

PplJDHirapara commented on src/app/recipes/recipe.service.ts:
```diff
@@ -0,0 +1,49 @@
+    recipeSelected = new EventEmitter<Recipe>();
+
+    private recipes: Recipe[] = [
+        new Recipe(
```
Take the mock data from the mock file. You can have a mock data file under the resources or asset from where you can pick this values

PplJDHirapara commented on src/app/recipes/recipe.service.ts:
```diff
@@ -0,0 +1,49 @@
+    };
+
+    getRecipe(index: number) {
+        return this.recipes.slice()[index];
```
Can't you return directly the element of that index instead of doing the slice?

PplPrinceDalsaniya replied:
Yes we can Directly return the elements. But that would return the same instance which is stored here. So, if we do 'slice()' then it will return a copy of the original array. This way if we if any unwanted changes happen to array then it wont affect the original array.

PplJDHirapara commented on src/app/recipes/recipes-detail/recipes-detail.component.ts:
```diff
@@ -0,0 +1,32 @@
+})
+export class RecipesDetailComponent implements OnInit {
+  recipe: Recipe;
+  id: number;
```
Update the name from id to some descriptive that for what it is used

PplJDHirapara commented on src/app/recipes/recipes-list/recipe-item/recipe-item.component.ts:
```diff
@@ -0,0 +1,15 @@
+})
+export class RecipeItemComponent implements OnInit {
+  @Input() recipe: Recipe;
+  @Input() index: number; 
```
Update index name to meaningfull variable name

PplJDHirapara commented on src/app/shared/ingredient.model.ts:
```diff
@@ -0,0 +1,11 @@
+    // public name : string;
+    // public quentity : number;
+
+    // constructor(name: string, quentity: number) {
```
Remove commented code

PplPrinceDalsaniya replied:
This was for my reference, that we can do it both ways. 
Commented code is long way to do the same thing done bellow.

Now I have it in NotionNotes. So, I will remove all the commented codes.

PplPrinceDalsaniya committed:
Replace EventEmitters with subjects

PplPrinceDalsaniya committed:
TD: adding shopping list form

- Added ngSubmit and ngForm on the form.
- ngModel in all input fields
- removed onClick evelnt from button.

PplPrinceDalsaniya committed:
Adding validations to add item form

- made name and quentity required.
= added validation for adding only positive integers in quentity,
- disable the add button if any of them is not valid.

PplPrinceDalsaniya committed:
Allowing the selection of items in shopping list.

PplPrinceDalsaniya committed:
Loading the shopping list item into the form

PplPrinceDalsaniya committed:
Updating Existing Ingredients

PplPrinceDalsaniya committed:
Resetting the form after adding ir updating.

- Also we need to stop edit mode once item is updated.

PplPrinceDalsaniya committed:
Added code for clearForm button to work.

PplPrinceDalsaniya committed:
Allowing Deleting of shopping list ingredient.

- show delete button only if in edit mode.

PplPrinceDalsaniya committed:
Creating template for Recipe Edit Form

PplPrinceDalsaniya committed:
Creating form for editing recipes

PplPrinceDalsaniya committed:
Sync recipe edit form with the foem we created.

PplPrinceDalsaniya committed:
Adding Ingredient controls to a Form Array

PplPrinceDalsaniya committed:
Adding new Ingredients control

PplPrinceDalsaniya committed:
validating user inputs

PplPrinceDalsaniya committed:
Submitting Recipe edit form

- it adds new recipe
- it updates the existing recipe

PplPrinceDalsaniya committed:
Adding delete and clear functionality

- from recipedetail page in dropdown menu, delete recipe working.
- while editing recipe, cancel button will take back to the detail page.
- if you save the changes in recipe it will take back to recipe detail page.

PplPrinceDalsaniya committed:
redirecting user to recipe list page after deleting the recipe

PplPrinceDalsaniya committed:
Adding Image preview

PplPrinceDalsaniya committed:
Deleting ingredients

- also added recipe service in app.module
  to avoid the error hapening because of it.

PplPrinceDalsaniya committed:
SettingUp DataStorage Service

- Created Data-storage Service
  - Injected HttpClient service into it.
- Imported it into appModule.

PplPrinceDalsaniya committed:
Storing Recipes
- DataStorageService
  - Injected RecipeService to get all current recipes into this service.
  - storeRecipe method simply puts all the recipes we currently have.
  - we subscribe to response firebase sends and prints it.
- Header Component
  - Added a click listener on "Save Data" button.
  - Onclick we call storeRecipes from dataStorageService.

PplPrinceDalsaniya committed:
Fetching Recipes
- dataStorageService
  - added fetchRecipes method which does http get request to firebase.
  - we added dataType that we are getting back in get request by <>.
  - we are calling setRecipes from recipeService while we fetch recipes.
- RecipeService
  - Added setRecipes method.
- Header component
  - added click listener to fetchData button.
  - calling fetchData method on click event.

PplPrinceDalsaniya committed:
Transforming ResponseData
- recipeService
  - Removed the dummy recipes array. and initialized it with empty one.
- DataStorageSErvice
  - added pipe after get response for adding ingredients property.
  - sometimes we save recipe without adding ingredient
  - then, firebase won't return ingredients property in response.
  - so, stopping our app to break due to type confilct, we check and add ingredients using pipe.

PplPrinceDalsaniya committed:
Resolving Data before Loading
- recipeResolverService
  - We have resolve method where we are fetching all recipes.
- AppRoutingModule
  - adding resolve property.

PplPrinceDalsaniya committed:
Little Bug on Resolver
- We only need to use reolver to fetch recipes if they are not already fetched.

PplPrinceDalsaniya committed:
Adding Auth Page
- Added auth component.
- Added it in declarations.
- added route for it.
- Added one link for this in headere component.

PplPrinceDalsaniya committed:
Switching between both Modes
- added one loggedIn flag for this.
- toggling button according to login status in template.

PplPrinceDalsaniya committed:
Handeling Form Input
- added validations for form.
- email for email and minLength for password
- disabled login button if validations are not true

PplPrinceDalsaniya committed:
Preparing SignUp Request

PplPrinceDalsaniya committed:
Sending a SignUp request

PplPrinceDalsaniya committed:
Added Error Handeling and Loading Spinner

PplPrinceDalsaniya committed:
Sending Login Requests and Handling errors

PplPrinceDalsaniya committed:
Creating and Sroring UserData
- created user model containing
  - email, id, token, tokenExpiration
- tapping into the response while successfully login or sugnup
  - creating new user in out local.

PplPrinceDalsaniya committed:
Doing all changes related to AuthState in the UI

PplPrinceDalsaniya committed:
Adding token to outgoing get request.

PplPrinceDalsaniya committed:
Attaching a token with Interceptor.

PplPrinceDalsaniya committed:
Adding logout functionality

PplPrinceDalsaniya committed:
Adding Auto Login
- Storing the token in localstorage
- checking if we have an valid user or not.
- if we have valid user then we can login them,

PplPrinceDalsaniya committed:
Adding Auto-Logout funcionality

PplPrinceDalsaniya committed:
Adding Auth Guard

PplPrinceDalsaniya committed:
Adding Alert Model Component

PplPrinceDalsaniya committed:
Using ngIf

PplPrinceDalsaniya committed:
Preparing Programmatic Creation

PplPrinceDalsaniya committed:
Creating Component Programatically
